### ISSUE TITLE
API: configurable ACLs

### ISSUE THREAD

**Issue by jeffomatic** [chain/chain#829]
**Grants** map a **guard** to a **policy**. Guards and policies may be associated via grants in a many-to-many fashion.

## Guards

Guards are effectively predicates which accept credentials as inputs and define the properties of the credentials which need to hold for an access control decision to be authorized/discharged.

For the purposes of ACL grants, guards are specified with **type** and a block of **data**.

Currently, there are three types of guards:

- `access_token`
- `x509`
- `localhost`

##### `access_token`

Guards of this type validate the presence of an access token. Data for this guard is an object specifying a token ID, e.g.:

```
{
  "id": "abc123"
}
```

##### `x509`

Guards of this type validate the presence of a client certificate. Data for this guard is an object specifying the fields and sub-fields present in the certificate. For example, the following will match certificates that include specific values for the Common Name and Organizational Unit sub-fields of the Subject field:

```
{
  "subject": {
    "cn": "Alice",
    "ou": "Engineering"
  }
}
```

##### `localhost`

Guards of this type will permit requests made from the loopback device. Grants on `localhost` do not need to be exposed by the grant/revoke API--instead, this is controlled via a special compile flag which enables or disables across-the-board RPC access to requests that satisfy the `localhost` guard.

## Policies

Initially, the following policies are available:

- `client-readwrite`: full access to the Client API
- `client-readonly`: access to read-only Client endpoints
- `network`: access to the Network API
- `monitoring`: access to monitoring-specific endpoints
- `internal`: access to multi-process synchronization endpoints (Raft, etc.)

## Grants

```
{
  "guard_type": <"access_token"|"x509">,
  "guard_data": <object>,
  "policy": <string>
}
```

The 3-tuple of values that compose a grant should be unique.

### List grants

Returns a list object containing all grants.

#### Request

Results are ordered by grant date in reverse chronological order.

```
POST /list-acl-grants
```

#### Response

```
{
  "items": [
    {
      "guard_type": <"access_token"|"x509">,
      "guard_data": <object>,
      "policy": <string>
    },
    ...
  ]
}
```

#### Request

```
POST /acl-grant

{
  "guard_type": <"access_token"|"x509">,
  "guard_data": <object>,
  "policy": <string>
}
```

#### Response

```
{
  "message": "ok"
}
```

### Revoke grant

Revokes a policy for the given guard. The special policy name `*` means that all policies should be revoked for the given guard.

```
POST /acl-revoke

{
  "guard_type": <"access_token"|"x509">,
  "guard_data": <object>,
  "policy": <string>
}
```

#### Response

```
{
  "message": "ok"
}
```

## Updates to access token API

The current API allows you to create both network and client access tokens. Since all access tokens share the same namespace, we can think of these as specializations of a more general access token object that can be bound to any of the available policies.

#### Request

The `type` parameter is now **deprecated**. Clients should create tokens and then use the ACL entry API for assigning a policy to the new token.

If the `type` parameter is blank, a new access token will be created with no associated grants.

For backward compatibility, using the deprecated `type` field will create a token and grant policies as follows:

`type` | Policy
--- | ---
`client` | `client-readwrite`
`network` | `network`

#### Request

The `type` parameter for this request is now **deprecated**.

#### Response

The `type` field in the token response object is now **deprecated**. For backward compatibilty, the value of this field should be based on the token's associated policies according to the following table:

`policies` | `type`
--- | ---
`client-readwrite` only | `client`
`network` only | `network`
anything else | `null`

### create-token

The correct usage for `corectl create-token` is as follows:

```
# Access tokens
corectl create-access-token <name> # Generate a new token
corectl acl-grant access_token <name> <policy1> <policy2> ... # Grant policies for the token

# X509
corectl acl-grant x509 '{"subject":{"cn": "...", "ou": ...}}' <policy1> <policy2>
```

The following patterns are **deprecated**:

```
corectl create-token [name] # creates a client-readwrite token
corectl create-token -net [name] # creates a network token
```

**Comment by jeffomatic** [comment]
@tarcieri recommends this as a reference point: https://github.com/square/rails-auth/wiki/Access-Control-Lists

**Comment by tarcieri** [comment]
I would recommend a data model that looks a bit more like this (pseudocoded in YAML):

```yaml
acl:
  policy: # acl has_many policies
    name: "core"
    resources: # policy has_many resources
      - path: "/*" # Would suggest avoiding regex unless you think you absolutely need it
        method: ALL # ALL means all *known* methods, not a wildcard
    allow_any_of: # policy has_many guards
      x509:
        cn: "core"
        ou: "production"
      token:
        type: "break-glass"
```

**Comment by jeffomatic** [comment]
@tarcieri Thanks. A couple of questions:

1. Where do specific credential identifiers fit in? The design in the parent comment puts the credential at the top level of the ACL entry, mapping credentials (either a cert subject or a token ID) to a scope (which could abstractly represent a set of policies).
2. When you say "acl has_many policies", this means that an ACL _references_ policies, by some ID or name?

**Comment by jbowens** [comment]
Are access tokens still a preferred authentication credential? With TLS everywhere, I'd hoped we could consolidate and deprecate the access tokens.

**Comment by jeffomatic** [comment]
> Are access tokens still a preferred authentication credential? With TLS everywhere, I'd hoped we could consolidate and deprecate the access tokens.

This is a broader long-term product question, but I don't think we can ditch them very quickly even if we decide to. In any case, I'm hoping that the presence of tokens doesn't provide burdensome constraints for our ACL design.

**Comment by jbowens** [comment]
>  In any case, I'm hoping that the presence of tokens doesn't provide burdensome constraints for our ACL design.

Their absence would allow ACLs like @tarcieri suggested up above. In his example, all certs with the "core" Common Name and "production" Organizational Unit signed by a trusted CA would be given the specified access.

I think we could achieve the same thing for access tokens by introducing "types" of access tokens, so an individual access token maps to a type. That access corresponding to that type of access token is specified in the ACL.

**Comment by tonychain** [comment]
> Where do specific credential identifiers fit in? The design in the parent comment puts the credential at the top level of the ACL entry, mapping credentials (either a cert subject or a token ID) to a scope (which could abstractly represent a set of policies).

Making an AuthZ decision about credentials is performed by a guard (in the `allow_any_of` section of the ACL).

Ideally these are simple, pluggable, and easy-to-extend. They can also be relatively self-contained. Here's an example from rails-auth for handling X.509:

https://github.com/square/rails-auth/blob/master/lib/rails/auth/x509/matcher.rb

> When you say "acl has_many policies", this means that an ACL references policies, by some ID or name?

I just mean the application-wide ACL logically consists of one-or-more policies. Oh, and since I didn't explicitly state it before: the default should be DENY.

> Are access tokens still a preferred authentication credential? With TLS everywhere, I'd hoped we could consolidate and deprecate the access tokens.

The ACL format I've described allows a mix-and-match approach. We used this at Square to e.g. allow SSO tokens in lieu of X.509 certificates.

Re: bearer credentials in general, I think they're useful, but ideally each one has an associated, revokable session. I would like to say monitoring systems are a nice use case, but one thing it's nice to have your monitoring system detect is if a e.g. a cert expires and you can no longer contact the service. I still think it's useful to allow bearer credentials to grant access to arbitrary parts of the system though.

It's probably worth keeping whatever we have now around until we have a better bearer credential system to replace it.

> I think we could achieve the same thing for access tokens by introducing "types" of access tokens, so an individual access token maps to a type. The access corresponding to that type of access token is specified in the ACL.

This can be accomplished relatively simply by attaching a list of permissions to each token, authenticated with HMAC.

**Comment by jeffomatic** [comment]
@kr Updated the parent based to today's conversation. Couple of caveats:

- I went ahead with the many-to-many relationship between credentials and policies. It was easy enough to express this up front, and we can skip a deprecation cycle if we build this in up front.
- I'm not sure if "ACL entry" is the appropriate name for the object that defines authz relationships between creds and policies.

**Comment by tarcieri** [comment]
Entry sounds OK. Otherwise I'd go with "guard"

**Comment by kr** [comment]
This latest version looks really awesome to me. I have a small suggested revision: we should probably rename the `cred_type` and `cred_id` fields.

We will probably start off with three (not two) guard types: `access_token`, `x509`, and `localhost`. In the future we might have more, including something that might require an access_token *and* a TLS cert (as an example that @tonychain brought up at one point). So we don't have to change any semantics right now, but it would awkward in that scenario for this field to still be called `cred_type`. Maybe we should call it `guard_type`.

Then, `cred_id` isn't really an ID, for example for an x509 cert it could be a subject name with some of the fields replace by wildcards, like `OU=*/CN=*.example.com`. In general, the interpretation of that field is determined by the guard type or credential type. So I'd suggest calling it `guard_data`.

**Comment by jeffomatic** [comment]
>  In the future we might have more

I'm down to future-proof, as long as we feel like this is the likely future scenario. I do feel like "guard" is quite a bit less legible than "credential" in the short-term. One thing that would help me out is a one-line definition of "guard"...we've been using it and I only have a hand-wavy sense of what it means.

**Comment by tarcieri** [comment]
@jeffomatic 

Guards are effectively predicates which accept credentials as inputs and define the properties of the credentials they wish to examine for an access control decision to be authorized/discharged.

https://www.cs.cornell.edu/fbs/publications/chptr.CredsBased.pdf

> Credentials-based authorization uses credentials and guards to enforce intensionally-defined authorization policies. Credentials convey beliefs about principals and/or the system state; guards employ logical inference to grant requests only when some specified goal formula is shown to hold. The goal formula is a logical formula involving (i) predicates characterizing beliefs that principals hold and (ii) predicates characterizing which sources of credentials the guard trusts
to convey accurate beliefs about one or another aspect of reality.

**Comment by jeffomatic** [comment]
Cool, "guards as predicates" is a great verbalization of what was floating in my head. Okay, I'm down.

**Comment by jeffomatic** [comment]
Continuing down the logic of "guards as predicates", is `guard_params` a more precise name than `guard_data`?

**Comment by kr** [comment]
> Continuing down the logic of "guards as predicates", is `guard_params` a more precise name than `guard_data`?

Makes no difference to me.

**Comment by jeffomatic** [comment]
@kr Now that future extensibility is on the table, does it make sense to make `guard_data` deserialize to a `map[string]interface{}`, rather than `string`?

**Comment by kr** [comment]
> does it make sense to make `guard_data` deserialize to a `map[string]interface{}`, rather than `string`?

I can argue this either way. Really not sure. My instinct is to keep it as a string for simplicity and leave it totally up to the guard to interpret (maybe parse it as json, or maybe just treat it as the name of something). But I don't feel super strongly about it.

**Comment by tarcieri** [comment]
@jeffomatic I think there's a lot more flexibility possible if you use a map instead of just a string. It also makes it easier to specify the properties you care about, especially with X.509 as Distinguished Names are effectively name-value pairs.

**Comment by jeffomatic** [comment]
I think I'm starting to lean toward Team Map, unless we can make the case that the guard will never need more than one piece of specifying information (in the form of the string).

**Comment by jeffomatic** [comment]
The one argument against `guard_data` as a map is canonicalization, which we may be able to file away as an implementation detail. I would want the following to be treated as describing the same guard:

```
A:
{
  "guard_type": "foo",
  "guard_data": {
    "hello": "world",
    "ooga": "booga"
  }
}

B:
{
  "guard_type": "foo",
  "guard_data": {
    "ooga": "booga",
    "hello": "world"
  }
}
```

In particular, I wouldn't want to be able to grant the same policy set to both A and B above, as the grant should be unique. I would also want the grants to sort easily based on the contents of `guard_data`. Maybe this is not a big deal?

**Comment by kr** [comment]
If there were some standardized and widely-supported string format for a subject name (or distinguished name), I might argue for using that and keeping it as a string.

Maybe something like https://tech.labs.oliverwyman.com/blog/2007/10/30/whats-in-a-distinguished-name/ but honestly ¯\\\_(ツ)_/¯

**Comment by kr** [comment]
> canonicalization, which we may be able to file away as an implementation detail

Luckily package json always sorts map keys when marshaling, so I think we are pretty much covered there.

**Comment by kr** [comment]
To operate on distinguished names, we will probably be working with these suckers: https://godoc.org/crypto/x509/pkix#Name.

**Comment by tarcieri** [comment]
> If there were some standardized and widely-supported string format for a subject name (or distinguished name), I might argue for using that and keeping it as a string.

There are pretty much two: X.500 Distinguished Names, which are "little endian" (i.e. the most specific part comes first), and "DirNames", a "big endian" counterpart which looks closer to a typical directory hierarchy. In either case, they describe a set of name/value pairs.

**Comment by boymanjor** [comment]
I was thinking about @kr's thoughts on `corectl` being a complete client to `cored` and not bypassing authn/z. The stated problem is that we would need a way to setup initial auth without setting the `loopback_auth` flag. What if we introduced a fourth guard type (`unconfigured`) which was inferred from the conditions of a request: 
- requested from localhost
- the `cored` has not been configued

We could also introduce a new policy type (`init`) that allowed access to endpoints that are associated with initializing a core: 
- creating access tokens and acl-grants
- creating block key's for the mockhsm

Essentially, all the functionality of `corectl` would be represented. This would allow us to get rid of the database dependency of `corectl` and have it completely adhere to the ACLs as a client.

I believe we could also get rid of the migration functionality in `corectl`. At startup `cored` will run any necessary migrations before listening for requests.

**Comment by kr** [comment]
I'd really like to solve this in a way that introduces as few special cases as possible. Ideally none. It's pretty important to keep this system simple so that the user's mental model has a better chance of being accurate. It's hard enough to configure ACLs correctly even without incidental complexity.

Luckily, I think we can.

Forgetting about corectl for a moment, let's look at cored. We will need a way to set up cored so that it can communicate with its other processes in the same cluster. In practice, this means giving each cored a TLS certificate and private key that it'll use to authenticate itself as a server (and also as a client) when another cored connects. For the purpose of authn, cored doesn't really care which other cored it's talking to; they are all equivalent. So they'll all have the same or closely related Subject Names in their certs. (They might even have a copy of the same cert and key, but whether they do or not makes no difference for our purposes here.)

So each cored has its own cert and knows its own Subject Name. It also knows that the rest of the cluster uses the same Subject Name. So it should have a built-in authorization policy granting this name (which it can read from its own cert) `internal` access. This built-in rule could be merely generated on first setup, or it could be hard-coded in a way that can't be deleted. Either way would help here. On the local unix system, this info needs to be protected in a way that lets only Chain Core access the private key. Something like unix file permissions.

Now let's look at corectl on that same machine. We know that the cored has a particular Subject Name, and that it allows any client authenticated with that SN access to the `internal` policy. We also know that there's a private key sitting on the local filesystem that can be used in exactly that way.

TL;DR:

- put cored's TLS cert+key in `CORED_DATA_DIR`
- make cored read them from there
- have corectl *also* read them from there, and use them to authenticate itself to cored locally
- the files perms will let the `chaincore` unix user, and not local randos, read them

**Comment by jeffomatic** [comment]
Sorry @kr I flubbed on the links and made you comment on the wrong issue. Here's your original comment: https://github.com/chain/chain/issues/857#issuecomment-294560762

> > `corectl acl-grant x509 '{"subject":{"cn": "...", "ou": ...}}' `
>
> Now that I see this written out, I realize it would be super nice to be able to take e.g. the output of openssl and copy-paste it to the CLI. Which would imply parsing some standard notation for Subject Names. I would suggest not doing that now, but I think it would be a substantial UX improvement.

Agree. I bet we could make corectl (if not the API as well) permissive, since I'm betting the JSON serializations and OpenSSL serializations are mutually exclusive.

**Comment by hsluoyz** [comment]
Hi, I'm the author of [casbin](https://github.com/hsluoyz/casbin). It is an authorization library that supports models like ACL, RBAC, ABAC.

```
{
  "guard_type": <"access_token"|"x509">,
  "guard_data": <object>,
  "policy": <string>
}
```
Basically, I think this granting can be expressed by casbin rules. I don't know what guard_data is, but we should use an ID of it here. Putting a large amount of data into access control rules is not a good idea. Let me know if there's any question:)

**Comment by jeffomatic** [comment]
Implementation work here is complete. Still TODO:

- SDK work
- Conversation about naming
